packages/upload-core/src/upload.js

```diff
@@ -334,11 +335,47 @@ class Upload {
     return true;
   }
 
+
+  async recursingUncorruptChecker(obj) {
+    await new Promise((resolve,reject) => {
+      setTimeout(() => {
+        resolve();
+        reject();
+      }, 100)
+    });
+    if (obj.isUncorrupt !== -1 ) {
+      return obj.isUncorrupt === 1;
+    }
+    return this.recursingUncorruptChecker(obj);
+  }
+
+  isUncorrupt() {
+    const obj = {isUncorrupt : -1};
+    if (this.file.type === 'image/tiff') {
+      const check =  (event: ProgressEvent<FileReader>) => {
+          try {
+            Tiff.pageCount(event.currentTarget.result);
+            obj.isUncorrupt = 1;
+          } catch (error) {
+            console.log(error);
+            this.setError('rejected', 'The tiff file is corrupt.');
+            obj.isUncorrupt = 0;
+          }
+      };
+      const fileReader = new FileReader()
+      fileReader.addEventListener("loadend", check);
+      fileReader.readAsArrayBuffer(this.file);
+      return this.recursingUncorruptChecker(obj);
+    }
+    return true;
+  }
+
   isValidFile() {
     return (
       this.isAllowedFileNameCharacters() &&
       this.isAllowedFileTypes() &&
-      this.isValidSize()
+      this.isValidSize() &&
+      this.isUncorrupt()
```

Comment:
It seems like the `isUncorrupt` function is for a specific type of use case that needs to run before an upload starts. I think instead of adding `isUncorrupt` to the general `upload` class, we should add a way for user to pass in "preStart" callbacks that execute before the actual upload begins instead. Something like

```js
start() {
    if (!this.isValidFile()) {
      return;
    }

    // this should probably get set in constructor
	let preStartValidated = true 

    // if cb condition fails, set this.preStartValidated
    for (const cb of props.onPreStart)  cb(); 
	if (!this.preStartValidated) {
      return;
    }

    this.upload.start();
}
```

This way you are still able to add the functionality you need, and the class opens up functionality for others if needed

Reply:
sure I can add that feature.

Reply:
So, if I leave the code as is in upload-core, would it be alright if I create a new PR for that onFilePreUpload callback for the upload component in the availity-react repo?

Reply:
@dapatiga Note that what I posted above would break down if the first callback fails but the last callback succeeds. So we would need a way to exit the callbacks early and return as soon as one callback fails validation. Or another option could be something like
```js
// this should probably get set in constructor
let preStartValidationResults = []; 

// if cb condition fails, push failure onto array
for (const cb of props.onPreStart){
	const validationResult = cb(); 
	this.preStartValidationResults.push(validationResult); // some T/F value
}

if (this.preStartValidated.some((result) => !result)) // return if validation result false

this.upload.start();
```

Reply:
Hey Chris, I modified it a bit.  Let me know what you think.  
I added the onPreStart method to be run in the constructor at the end.
``` js
    if (this.options.onPreStart) {
      this.options.onPreStart(this);
    }
```

This way, the developer can run setError before he calls start(), and the file and the name onPreStart actually means what it says.  

I also modified the start function to include a check for error messages.
``` js
start() {
    if (!this.isValidFile()) {
      return;
    } 
    if (this.errorMessage || this.error || this.status === 'rejected') {
      return;
    }
    this.upload.start();
  }
```

Reply:
> // this should probably get set in constructor
> let preStartValidationResults = []; 

I think this could be implemented in a single method, which might be overly simple, but I can go with this option if you insist.